jenkins-jobs/mci/project.rb

```diff
@@ -0,0 +1,138 @@
+
+    dependees = project.dependees.collect do |d|
+      basename(distribution, type, d.component, d.name)
+    end
+
+    publisher_dependees = project.dependees.collect do |d|
+      "#{basename(distribution, type, d.component, d.name)}_src"
+    end.compact
+    sourcer = SourcerJob.new(basename,
+                             type: type,
+                             distribution: distribution,
+                             project: project)
+    publisher = PublisherJob.new(basename,
+                                 type: type,
+                                 distribution: distribution,
+                                 dependees: publisher_dependees,
+                                 component: project.component,
+                                 upload_map: nil,
+                                 architectures: architectures)
+    binariers = architectures.collect do |architecture|
+      BinarierJob.new(basename, type: type, distribution: distribution,
+                                architecture: architecture)
+    end
+    jobs = [sourcer, binariers, publisher]
+    basename1 = jobs[0].job_name.rpartition('_')[0]
+    unless basename == basename1
+      raise "unexpected basename diff #{basename} v #{basename1}"
+    end
+
+    jobs << new(basename, project: project, jobs: jobs, dependees: dependees)
+    # The actual jobs array cannot be nested, so flatten it out.
+    jobs.flatten
+  end
+
+  # @! attribute [r] jobs
+  #   @return [Array<String>] jobs invoked as part of the multi-phases
+  attr_reader :jobs
+
+  # @! attribute [r] dependees
+  #   @return [Array<String>] name of jobs depending on this job
+  attr_reader :dependees
+
+  # @! attribute [r] project
+  #   @return [Project] project instance of this job
+  attr_reader :project
+
+  # @! attribute [r] upstream_scm
+  #   @return [CI::UpstreamSCM] upstream scm instance of this job_name
+  # FIXME: this is a compat thingy for sourcer (see render method)
+  attr_reader :upstream_scm
+
+  private
+
+  def initialize(basename, project:, jobs:, dependees: [])
+    super(basename, 'project.xml.erb')
+
+    # We use nested jobs for phases with multiple jobs, we need to aggregate
+    # them appropriately.
+    job_names = jobs.collect do |job|
+      next job.collect(&:job_name) if job.is_a?(Array)
+      job.job_name
+    end
+
+    @nested_jobs = job_names.freeze
+    @jobs = job_names.flatten.freeze
+    @dependees = dependees.freeze
+    @project = project.freeze
+  end
+
+  def render_phases
+    ret = ''
+    @nested_jobs.each_with_index do |job, i|
+      ret += MultiJobPhase.new(phase_name: "Phase#{i}",
+                               phased_jobs: [job].flatten).render_template
+    end
+    ret
+  end
+
+  def render_packaging_scm
+    PackagingSCMTemplate.new(scm: @project.packaging_scm).render_template
+  end
+
+  def render_upstream_scm
+    @upstream_scm = @project.upstream_scm # FIXME: compat assignment
```

Comment:
Get rid of all @upstream_scm mentions to drop compat which wouldn't affect mci.

Reply:
Turns out `@upstream_scm` is actually used in MCI sourcer jobs, I will kill it off in separate commit/PR.